src/plugins/vis_builder/public/application/components/data_tab/index.tsx

```diff
@@ -3,19 +3,336 @@
+
+  const indexField = isCountField
+    ? { type: 'count' }
+    : getIndexPatternField(fieldName, aggProps.indexPattern?.fields ?? []);
+
+  if (!indexField) return [];
+
+  // Get all aggTypes allowed by the schema and get a list of all the aggTypes that the dragged index field can use
+  const aggTypes = aggService.types.getAll();
+  // `types` can be either a Bucket or Metric aggType, but both types have the name property.
+  const allowedAggTypes = filterByName(
+    aggTypes[destinationSchema.group] as BucketAggType[],
+    destinationSchema.aggFilter
+  );
+
+  return (
+    allowedAggTypes
+      .filter((aggType) => {
+        const allowedFieldTypes = aggType.paramByName('field')?.filterFieldTypes;
+        return filterByType([indexField], allowedFieldTypes).length !== 0;
+      })
+      .filter((aggType) => (isCountField ? true : aggType.name !== 'count'))
+      // `types` can be either a Bucket or Metric aggType, but both types have the name property.
+      .map((aggType) => (aggType as BucketAggType).name)
+  );
+};
+
+export const createNewAggConfig = ({
+  fieldName,
+  sourceGroup,
+  destinationSchema,
+  aggProps,
+  aggService,
+  sourceAgg,
+}) => {
+  const schemaAggTypes = (destinationSchema.defaults as any).aggTypes;
+
+  const destinationValidAggType = getValidAggTypes({
+    fieldName,
+    sourceGroup,
+    destinationSchema,
+    aggProps,
+    aggService,
+    sourceAgg,
+  });
+
+  const allowedAggTypes = schemaAggTypes
+    ? schemaAggTypes.filter((type: string) => destinationValidAggType.includes(type))
+    : [];
+
+  aggProps.aggConfigs?.createAggConfig({
+    type: allowedAggTypes[0] || destinationValidAggType[0],
+    schema: destinationSchema.name,
+    params: {
+      field: fieldName,
+    },
+  });
+};
+
 export const DataTab = () => {
+  // Field Selector panel States
+  const fieldSelectorGroups = [
+    'preDefinedCountMetric',
+    'categoricalFields',
+    'numericalFields',
+    'metaFields',
+  ];
+
+  // Config panel States
+  const vizType = useVisualizationType();
+  const editingState = useTypedSelector(
+    (state) => state.visualization.activeVisualization?.draftAgg
+  );
+  const schemas = vizType.ui.containerConfig.data.schemas;
+  const {
+    services: {
+      data: {
+        search: { aggs: aggService },
+      },
+    },
+  } = useOpenSearchDashboards<VisBuilderServices>();
+
+  const aggProps = useAggs();
+  const [schemaDisplayStates, setSchemaDisplayStates] = useState(() => {
+    return schemas.all.reduce((acc, schema) => {
+      acc[schema.name] = [];
+      return acc;
+    }, {});
+  });
+  const dispatch = useTypedDispatch();
+
+  useEffect(() => {
+    const newState = schemas.all.reduce((acc, schema) => {
+      acc[schema.name] = [];
+      return acc;
+    }, {});
+    const upDatedState = { ...newState, ...schemaDisplayStates };
+    setSchemaDisplayStates(upDatedState);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
why we need this?

Reply:
When user changes the chart type from the right pane (i.e from bar to line or metric etc.) with already selected configurations we want these states to persist for the ones that we can carry over. To account for this use case I have added an useEffect hook that tries to carryover the older config states that we already had.